### ISSUE TITLE
USB problems with Retina MacBook Pro

### ISSUE THREAD

**Issue by kyleellman** [Roastero/Openroast#66]
Hi,

I originally installed Openroast on an old MacBook Pro (with the CD drive), and it worked perfectly.

But I recently tried installing Openroast on a newer Renita Macbook Pro (still with normal USB-A ports), and it doesn't really work well:
- In order to talk to the roaster at all, you must connect the roaster first, and then launch the application.
- While roasting with a recipe, the application seems to lose its connection with the roaster after a couple minutes, and even though the settings change in the app throughout the recipe, the roaster's settings doesn't change with the application, so it might be stuck at full heat, or no heat indefinitely.

Does anyone know what's going on here? Is it a driver issue?

Thanks.

**Comment by int3ll3ct** [comment]
I'm assuming you're running v1. 2,is that correct?

**Comment by kyleellman** [comment]
I'm using v1.2.1. Should I be using v1.2 instead?

**Comment by int3ll3ct** [comment]
No use the latest, v1.2.1 is fine. I am assuming that you installed the driver software on the new Mac, which is part of the overall package. If you hadn't, there would be zero comms with the hardware. I tried the app using a 2017 Pro (it only has USB C ports) and with a multiport hub, it worked. My theory is that you have one of these 2015-era "it doesn't work with my USB port" problems, which anecdotally seems to be associated with a certain era of USB 3 chipsets. However, your description does not match that behavior 100%. If you have a crusty USB hub handy, I would plug the device into that hub, and plug the hub into the Mac. This seems to work around the se types of issues. These basic communication problems are hard to debug from the python level (all this code is python) and finding root causes is unlikely. I'll see how easy it is to generate logs in the driver context, to give us some visibility.

**Comment by kyleellman** [comment]
I installed the drivers before installing the application as I did on the older 2012 MacBook Pro (where it worked), but the newer one I'm having trouble with is in fact an early 2015 Macbook Pro. I don't have a usb hub laying around, but I think I'll get a cheap one somewhere and see if that fixes the problem.

Thanks for the response!

**Comment by int3ll3ct** [comment]
Ok let's keep this open, and we will figure out what to do whether that
works or not. At minimum I'd want to get the specifics of the Mac model to
add to a list of known issues.

**Comment by kyleellman** [comment]
I got a USB hub and connected the roaster to my MacBook Pro through that. The roaster still needs to be connected prior to launching the application, but the hub seems to have solved the problem! I was able to run a full roasting recipe without any problems.

Just for data, this occurs on a `MacBook Pro (Retina, 13-inch, Early 2015)`.

Thanks!

**Comment by pollakben** [comment]
I am experiencing a very similar problem running the latest stable release on a MacBook Pro (Retina, 13-inch, Late 2013).  Openroast can find but cannot connect to my sr700 roaster.  It says it will take about 20 seconds but it does not actually finish

**Comment by int3ll3ct** [comment]
Yes, the service used to enumerate USB devices is distinct from the interface used to open the USB connection. I am assuming that you have installed the necessary USB driver for Mac. I'd say try the workaround suggested in this thread and see if that works.

**Comment by pollakben** [comment]
that fixed it! thanks for the speedy answer!

**Comment by pollakben** [comment]
I left town and came home to find the problem has started again, though no changes to my system have been made.

**Comment by gkalinin** [comment]
I'm having the same problem using a Macbook Pro Retina 13 inch early 2015 (running High Sierra) as well as a Macbook Air 13 inch 2017 (running Sierra) – i.e. they will connect if I launch Openroast v1.2.1 after the roaster is turned on and the usb cable connected and seems to function for the first couple of minutes but then looses the connection ... the temperature flatlines, the roaster goes to cooling mode and none of the buttons on the Openroast work anymore.   

I've ordered a USB hub and will let you know if that does the trick. Thanks!

**Comment by gkalinin** [comment]
Update: I tested using a USB hub (Anker 4-Port USB 3.0 Ultra Slim Data Hub) and everything seems to be working fine now.  Thank you!

**Comment by int3ll3ct** [comment]
Glad to hear it gkalinin.  Unfortunately I really do not know the root cause issue of this problem, I suspect it's related to the USB implementation in the roaster - it's a cheap clone chip of the industry-standard FTDI serial-usb converter, and either the electrical signaling is marginal, or the low-level USB code is not optimal.  I do not control how the SR700 is manufactured (I have no affiliation with the company), so we'll leave this technical exchange posted here for others to see, and in the future I'll roll these observations into the readme.  Thanks.

**Comment by balogms** [comment]
Hi, I'm having a similar problem. I loaded both the driver and the latest Openroast. The device connects, but then shutsdown immediately after I hit the roast button. Any thoughts?

**Comment by kyleellman** [comment]
If you connect the roaster through a USB hub instead of directly to your MacBook, it should roast just fine as long as you open the app after connecting. A bit of an odd workaround, but it works for me every time.

**Comment by balogms** [comment]
Thanks Kyle, were you connecting to a usb-c? I used an adapter to go from  usb to usb-c but I’ll try a hub.

**Comment by kyleellman** [comment]
Oh, I'm not sure. I have an older macbook pro with USB-A ports, so I can't try to replicate your exact issue. But I'd recommend passing through a hub if you have one for good measure if nothing else works.

**Comment by bdruth** [comment]
Was having this issue on my MBP 16" with a Hyper Drive USB Type-C 5-in-1 Hub. I grabbed my son's old MBP2012 and everything magically worked. FWIW, I also tried using it through Windows (VMware) and the same behavior persisted. Something's just not right with the USB control, probably the driver isn't implemented right somehow.
